private/buf/bufsync/bufsyncapi/sync_handler.go

```diff
@@ -0,0 +1,526 @@
+) bufsync.Handler {
+	return &syncHandler{
+		logger:                               logger,
+		container:                            container,
+		repo:                                 repo,
+		createWithVisibility:                 createWithVisibility,
+		moduleIdentityToRepositoryIDCache:    make(map[string]string),
+		moduleIdentityToDefaultBranchCache:   make(map[string]string),
+		syncServiceClientFactory:             syncServiceClientFactory,
+		referenceServiceClientFactory:        referenceServiceClientFactory,
+		repositoryServiceClientFactory:       repositoryServiceClientFactory,
+		repositoryBranchServiceClientFactory: repositoryBranchServiceClientFactory,
+		repositoryTagServiceClientFactory:    repositoryTagServiceClientFactory,
+		repositoryCommitServiceClientFactory: repositoryCommitServiceClientFactory,
+	}
+}
+
+func (h *syncHandler) ResolveSyncPoint(
+	ctx context.Context,
+	moduleIdentity bufmoduleref.ModuleIdentity,
+	branchName string,
+) (git.Hash, error) {
+	service := h.syncServiceClientFactory(moduleIdentity.Remote())
+	syncPoint, err := service.GetGitSyncPoint(ctx, connect.NewRequest(&registryv1alpha1.GetGitSyncPointRequest{
+		Owner:      moduleIdentity.Owner(),
+		Repository: moduleIdentity.Repository(),
+		Branch:     branchName,
+	}))
+	if err != nil {
+		if connect.CodeOf(err) == connect.CodeNotFound {
+			// No syncpoint
+			return nil, nil
+		}
+		return nil, fmt.Errorf("get git sync point: %w", err)
+	}
+	hash, err := git.NewHashFromHex(syncPoint.Msg.GetSyncPoint().GitCommitHash)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"invalid sync point from BSR %q: %w",
+			syncPoint.Msg.GetSyncPoint().GetGitCommitHash(),
+			err,
+		)
+	}
+	return hash, nil
+}
+
+func (h *syncHandler) IsGitCommitSynced(
+	ctx context.Context,
+	moduleIdentity bufmoduleref.ModuleIdentity,
+	hash git.Hash,
+) (bool, error) {
+	service := h.referenceServiceClientFactory(moduleIdentity.Remote())
+	res, err := service.GetReferenceByName(ctx, connect.NewRequest(&registryv1alpha1.GetReferenceByNameRequest{
+		Owner:          moduleIdentity.Owner(),
+		RepositoryName: moduleIdentity.Repository(),
+		Name:           hash.Hex(),
+	}))
+	if err != nil {
+		if connect.CodeOf(err) == connect.CodeNotFound {
+			// Repo is not created.
+			return false, nil
+		}
+		return false, fmt.Errorf("get reference by name: %w", err)
+	}
+	return res.Msg.Reference.GetVcsCommit() != nil, nil
+}
+
+func (h *syncHandler) IsGitCommitSyncedToBranch(
+	ctx context.Context,
+	moduleIdentity bufmoduleref.ModuleIdentity,
+	branchName string,
+	hash git.Hash,
+) (bool, error) {
+	repositoryID, err := h.getRepositoryID(ctx, moduleIdentity)
+	if err != nil {
+		return false, err
+	}
+	service := h.repositoryBranchServiceClientFactory(moduleIdentity.Remote())
+	res, err := service.ListRepositoryBranchesByReference(ctx, connect.NewRequest(&registryv1alpha1.ListRepositoryBranchesByReferenceRequest{
```

Comment:
Is this page-able? I guess we need to keep paging?

Reply:
You're right, I need to keep paging this.